### PR TITLE
Creatures model

### PR DESCRIPTION
**Pull request by vitsalis**

### CODE REVIEW

File: models/creature.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<?php
+    class Creature extends ActiveRecordBase {
+        public $id;
+        public $game;
```
Comment (dionyziz) [inline models/creature.php:3]: I don't see how Creature objects can be instantiated and has-a relationships be materialized.

File: models/creature.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        public $user;
+        public $round;
+        public $x;
```
Comment (dionyziz) [inline models/creature.php]: None of these attributes are being set, how can I use these objects?

File: models/game.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            }
+            if ( !is_int( $this->height ) ) {
+                throw new ModelValidationException( 'height_not_valid' );
+            }
+        }
+
+        protected function create() {
+            $width = $this->width;
+            $height = $this->height;
+            $created = $this->created;
```
Comment (dionyziz) [inline models/game.php]: Should default to current datetime.

File: models/game.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<?php
+    class Game extends ActiveRecordBase {
+        public $id;
+        public $created;
+        public $width;
+        public $height;
+        public $rounds;
```
Comment (dionyziz) [inline models/game.php:13]: This is not set.

File: models/game.php (hunk 3 above)
Comment (vitsalis) [inline models/game.php]: Should I do that in the model?

File: models/intent.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    define( 'DIRECTION_NORTH', 1 );
+    define( 'DIRECTION_EAST', 2 );
+    define( 'DIRECTION_SOUTH', 3 );
+    define( 'DIRECTION_WEST', 4 );
+    define( 'ACTION_ATACK', 0 );
+    define( 'ACTION_MOVE', 1 );
+    class Intent {
+        public $action;
+        public $direction;
+
```
Comment (dionyziz) [inline models/intent.php:67]: Intent belongs-to Creature, must keep reverse has-a relationship.

File: models/round.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
+<?php
+    class Round extends ActiveRecordBase {
+        public $creatures;
+        public $id;
+        public $game;
+
```
Comment (dionyziz) [inline models/round.php:6]: __construct?

File: models/game.php (hunk 4 above)
Comment (vitsalis) [inline models/game.php:13]: what do you mean it's not set?
should I have it like:
public $rounds = array(); ?

File: models/round.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        protected function create() {
+            $gameid = $this->game->id;
+            $roundid = $this->id;
+            foreach ( $this->creatures as $creature ) {
+                    $locationx = $creature->x;
+                    $locationy = $creature->y;
+                    $hp = $creature->hp;
+                    switch ( $creature->intent->direction ) {
+                        case 0:
```
Comment (dionyziz) [inline models/round.php]: Use dictionary.

File: test_game.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    $game->rounds = array();
+    for ( $i = 0; $i <= 2; ++$i ) {
+        $game->rounds[ $i ] = new Round();
+        $game->rounds[ $i ]->game = $game;
+        $game->rounds[ $i ]->id = $i; 
+        for ( $j = 0, $k = 0; $j <= 2; ++$j, $k += 2 ) {
+            if ( $i === 0 ) {
+                $game->rounds[ $i ]->creatures[ $j ] = new Creature(); 
+                $game->rounds[ $i ]->creatures[ $j ]->id = $j;
+                $game->rounds[ $i ]->creatures[ $j ]->game = $game;
```
Comment (dionyziz) [inline test_game.php]: Should be set automatically during creatures instantiation.

File: test_game.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                $game->rounds[ $i ]->creatures[ $j ]->x = 1;
+                $game->rounds[ $i ]->creatures[ $j ]->y = 0;
+                $game->rounds[ $i ]->creatures[ $j ]->hp = 10;
+                $game->rounds[ $i ]->creatures[ $j ]->intent = new Intent( 'ACTION_MOVE', 'DIRECTION_NONE' );
+                $game->rounds[ $i ]->creatures[ $j ]->save();
+            }
+            else {
+                $game->rounds[ $i ]->creatures[ $j ] = $game->rounds[ $i - 1 ]->creatures[ $j ];
+                switch ( $game->rounds[ $i ]->creatures[ $j ]->intent->direction ) {
+                    case 1:
```
Comment (dionyziz) [inline test_game.php]: Use constants.

File: test_game.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                        $game->rounds[ $i ]->creatures[ $j ]->y += 1;
```
Comment (dionyziz) [inline test_game.php]: Should not be part of unit test, write resolution in model and assert that it's correct in unit test.

File: test_game.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
+<?php
+    include 'models/dependencies.php';
+    include 'models/creature.php';
+    include 'models/round.php';
+    include 'models/intent.php';
+    include 'models/user.php';
+    include 'models/game.php';
+    $user = new User( 1 );
+    $game = new Game();
+    $game->created = date('Y-m-d H:i:s');
```
Comment (dionyziz) [inline test_game.php]: Must be set inside class during construction.

File: test_game.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    $game->width = $game->height = 10;
+    $game->save();
+    $game->rounds = array();
```
Comment (dionyziz) [inline test_game.php]: First round should be created automatically during game instantiation.

File: test_game.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    for ( $i = 0; $i <= 2; ++$i ) {
+        $game->rounds[ $i ] = new Round();
```
Comment (dionyziz) [inline test_game.php]: Next round must be created by calling ->nextRound.

File: test_game.php [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        $game->rounds[ $i ]->game = $game;
```
Comment (dionyziz) [inline test_game.php]: Belongs-to relationship must be assigned automatically, not part of API.

File: test_game.php [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        $game->rounds[ $i ]->id = $i;
```
Comment (dionyziz) [inline test_game.php]: Round id should be assigned automatically.

File: test_game.php [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+                $game->rounds[ $i ]->creatures[ $j ] = new Creature(); 
+                $game->rounds[ $i ]->creatures[ $j ]->id = $j;
+                $game->rounds[ $i ]->creatures[ $j ]->game = $game;
+                $game->rounds[ $i ]->creatures[ $j ]->user = $user;
+                $game->rounds[ $i ]->creatures[ $j ]->round = $game->rounds[ $i ]; 
+                $game->rounds[ $i ]->creatures[ $j ]->x = 1;
+                $game->rounds[ $i ]->creatures[ $j ]->y = 0;
+                $game->rounds[ $i ]->creatures[ $j ]->hp = 10;
+                $game->rounds[ $i ]->creatures[ $j ]->intent = new Intent( 'ACTION_MOVE', 'DIRECTION_NONE' );
+                $game->rounds[ $i ]->creatures[ $j ]->save();
```
Comment (dionyziz) [inline test_game.php]: You need to add some assertions, you're not testing anything other than the fact that the functions are callable and there are no run-time errors.

File: models/game.php (hunk 4 above)
Comment (dionyziz) [inline models/game.php:13]: No, set it to a collection containing the rounds of the current game (if it's not a new game). This API will be necessary for reading game history.

File: models/creature.php (hunk 1 above)
Comment (vitsalis) [inline models/creature.php:3]: I don't get what you mean

File: models/creature.php (hunk 2 above)
Comment (vitsalis) [inline models/creature.php]: they now are being set

File: test_game.php (hunk 10 above)
Comment (vitsalis) [inline test_game.php]: ok I'll do that.

File: test_game.php (hunk 9 above)
Comment (vitsalis) [inline test_game.php]: what do you mean?

File: test_game.php (hunk 16 above)
Comment (vitsalis) [inline test_game.php]: after each test, I check on database if the functions worked correctly.

File: test_game.php (hunk 8 above)
Comment (vitsalis) [inline test_game.php]: how?

File: test_game.php (hunk 15 above)
Comment (vitsalis) [inline test_game.php]: but it's the number of the round.. i.e: round 1 of the 3rd game => 
$game = new Game( 3 ); 
$game->round[ 1 ];

File: test_game.php (hunk 14 above)
Comment (vitsalis) [inline test_game.php]: how am I gonna do that?

File: models/round.php [hunk 17]
```diff
@@ -0,0 +1,84 @@
+<?php
+    class Round extends ActiveRecordBase {
+        public $creatures;
+        public $id;
+        public $game;
+
+        public function __construct( $game = false, $id = false ) {
+            if ( $id && $game ) {
```
Comment (dionyziz) [inline models/round.php]: if ( $id !== false && $game !== false )

File: models/round.php [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+            if ( !is_int( $this->game->id ) ) {
+                throw new ModelValidationException( 'gameid_not_valid' );
+            }
+        }
+
+        protected function create() {
+            $gameid = $this->game->id;
+            $roundid = $this->id;
+            foreach ( $this->creatures as $creature ) {
+                    $locationx = $creature->x;
```
Comment (dionyziz) [inline models/round.php]: Fix indentation

File: models/round.php [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+                    $locationy = $creature->y;
+                    $hp = $creature->hp;
+                    $direction = array(
+                        'NONE', 'NORTH', 'EAST', 'SOUTH', 'WEST'
```
Comment (dionyziz) [inline models/round.php]: Make this a map from constants to strings:

DIRECTION_NONE => 'NONE', etc.

This way we can change the value of the constants without changing this code.

File: test_game.php (hunk 9 above)
Comment (dionyziz) [inline test_game.php]: case DIRECTION_NORTH etc.

File: test_game.php (hunk 16 above)
Comment (dionyziz) [inline test_game.php]: Use assert to automate testing.

File: models/round.php [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+                            'hp', 
+                            'direction', 
+                            'action',
+                            'creatureid' 
+                        )
+                    );
+            }
+        }
+
+        public function nextCreature() {
```
Comment (dionyziz) [inline models/round.php]: Creatures must remain the same across rounds, e.g. if creature 4 belongs to user 25 during round 3, then creature 4 should also belong to user 25 during round 4. Not sure what this function is here for.

File: models/intent.php [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+    define( 'DIRECTION_WEST', 4 );
+    define( 'ACTION_NONE', 0 );
+    define( 'ACTION_MOVE', 1 );
+    define( 'ACTION_ATACK', 2 );
+    class Intent {
+        public $action;
+        public $direction;
+        public $creature;
+
+        public function __construct( $action, $direction ) {
```
Comment (dionyziz) [inline models/intent.php:68]: Pass constant value to model, not string. Conversion from string to constant value should be done at the controller level. The controller is the only place where we know what kinds of strings are accepted in URLs.

File: models/game.php [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+            }
+            if ( !is_int( $this->height ) ) {
+                throw new ModelValidationException( 'height_not_valid' );
+            }
+        }
+
+        protected function create() {
+            $width = $this->width;
+            $height = $this->height;
+            $created = $this->created = date('Y-m-d H:i:s');
```
Comment (dionyziz) [inline models/game.php]: style: spaces within ()

File: models/round.php (hunk 20 above)
Comment (vitsalis) [inline models/round.php]: I run this function only when I'm in round 0.

File: models/creature.php (hunk 1 above)
Comment (dionyziz) [inline models/creature.php:3]: This looks fine now.

File: models/round.php (hunk 20 above)
Comment (dionyziz) [inline models/round.php]: How are creatures associated with players?

Comment (vitsalis) [inline models/round.php]: a user has creatures

Comment (dionyziz) [inline models/round.php]: Where is $creature->user set when nextCreature() is called to create a new creature?

Comment (vitsalis) [inline models/round.php]: well we can have a parameter at nextCreature() about the user. Do you agree?

Comment (dionyziz) [inline models/round.php]: OK

File: models/intent.php [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+    define( 'DIRECTION_NONE', 0 );
+    define( 'DIRECTION_NORTH', 1 );
+    define( 'DIRECTION_EAST', 2 );
+    define( 'DIRECTION_SOUTH', 3 );
+    define( 'DIRECTION_WEST', 4 );
+    define( 'ACTION_NONE', 0 );
+    define( 'ACTION_MOVE', 1 );
+    define( 'ACTION_ATACK', 2 );
+
+    function convertDirection( $direction ) {
```
Comment (dionyziz) [inline models/intent.php]: These should be two different functions: directionStringToConst and directionConstToString. Similarly for Action.

File: test_game.php [hunk 24]
```diff
@@ -0,0 +1,47 @@
+<?php
```
Comment (dionyziz) [inline test_game.php]: Move this to unit test.

Comment (vitsalis) [inline test_game.php]: well you have to merge the testrun branch first, so I can merge unit testing here..
I have done a pull request.

File: models/creature.php [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+        public $user;
+        public $round;
+        public $locationx; 
+        public $locationy;
+        public $intent;
+        public $alive;
+        public $hp;
+        protected $gameid;
+        protected $userid;
+        protected $attributes = array( 'id', 'gameid', 'userid' );
```
Comment (dionyziz) [inline models/creature.php:13]: locationx, locationy, hp?

File: models/game.php [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+                    $this->rounds[ $i ] = new Round( $this, $i );
+                }
+            }
+            else {
+                $this->rounds = array();
+            }
+        }
+
+        protected function onBeforeCreate() {
+            $this->creaturesPerPlayer = rand( 100, 199 );
```
Comment (dionyziz) [inline models/game.php]: Make these constants configurable.

File: models/game.php [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+            $multiply = $this->creaturesPerPlayer * count( $this->users );
+            $this->width = rand( 3 * $multiply + 1, 4 * $multiply - 1 );
```
Comment (dionyziz) [inline models/game.php]: Make these bound constants configurable.

File: models/game.php [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+            $this->height = rand( 3 * $multiply + 1, 4 * $multiply - 1 );
+            $this->maxHp = rand( 100, 199 );
```
Comment (dionyziz) [inline models/game.php]: Make configurable.

File: models/game.php [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+            $this->created = date( 'Y-m-d H:i:s' );
+        }
+
+        public function genesis() {
+            $this->rounds[ 0 ] = new Round();
+            $id = 0;
+            for ( $i = 0; $i < count( $this->users ); ++$i ) {
```
Comment (dionyziz) [inline models/game.php]: foreach

File: models/round.php [hunk 30]
```diff
@@ -0,0 +1,72 @@
...
+            else {
+                $this->creatures = array();
+            }
+        }
+
+        protected function create() {
+            $gameid = $this->game->id;
+            $roundid = $this->id;
+            foreach ( $this->creatures as $creature ) {
+                    $locationx = $creature->locationx;
```
Comment (dionyziz) [inline models/round.php]: Indentation

File: test_game.php (hunk 24 above)
Comment (dionyziz) [inline test_game.php]: Please remove this file.

File: tests/models/game.php [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+    include_once 'models/user.php';
+
+    class GameTest extends UnitTestWithUser {
+        public function testInit() {
+            $game = new Game();
+            for ( $i = 1; $i <= 4; ++$i ) {
+                $game->users[] = $this->buildUser( $i );
+            }
+            $game->save();
+            $this->assertEquals( $game->width > 0, true, 'A game with users must have width' );
```
Comment (dionyziz) [inline tests/models/game.php]: assertTrue

File: tests/models/game.php [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+            $this->assertEquals( $game->height > 0, true, 'A game with users must have height' );
+            $this->assertEquals(
```
Comment (dionyziz) [inline tests/models/game.php]: assertTrue

File: tests/models/game.php [hunk 33]
```diff
@@ -0,0 +1,84 @@
...
+
+    class GameTest extends UnitTestWithUser {
+        public function testInit() {
+            $game = new Game();
+            for ( $i = 1; $i <= 4; ++$i ) {
+                $game->users[] = $this->buildUser( $i );
+            }
+            $game->save();
+            $this->assertEquals( $game->width > 0, true, 'A game with users must have width' );
+            $this->assertEquals( $game->height > 0, true, 'A game with users must have height' );
```
Comment (dionyziz) [inline tests/models/game.php]: assertTrue

File: models/round.php [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+                            'hp', 
+                            'direction', 
+                            'action',
+                            'creatureid' 
+                        )
+                    );
+            }
+        }
+
+        public function nextCreature( $user ) {
```
Comment (dionyziz) [inline models/round.php]: delete

File: tests/models/game.php [hunk 35]
```diff
@@ -0,0 +1,84 @@
...
+            // start from 1 because user's id starts from 1
+            for ( $i = 1; $i <= count( $game->users ); ++$i ) {
+                $userCountCreatures[ $i ] = 0;
+            }
+            $creatureCount = 0;
+            for ( $i = 0; $i < $game->width; ++$i ) {
+                for ( $j = 0; $j < $game->height; ++$j ) {
+                    if ( isset( $game->grid[ $i ][ $j ] ) ) {
+                        $creature = $game->grid[ $i ][ $j ];
+                        $userCountCreatures[ $creature->user->id ] += 1;
```
Comment (dionyziz) [inline tests/models/game.php]: ++$userCountCreatures...

File: tests/models/game.php [hunk 36]
```diff
@@ -0,0 +1,84 @@
...
+                        $this->assertTrue( $creature->locationx >= 0, "A creature's x coordinate must be positive" );
```
Comment (dionyziz) [inline tests/models/game.php]: :s/positive/non-negative

File: tests/models/game.php [hunk 37]
```diff
@@ -0,0 +1,84 @@
...
+                $creatureCount, 
+                'Each player must have a certain number of creatures' 
+            );
+            $creaturesPerUser = $creatureCount / count( $game->users );
+            foreach ( $userCountCreatures as $userCountCreature ) {
+                $this->assertEquals( $userCountCreature, $creaturesPerUser, 'Each user must have the same number of creatures' );
+            }
+            foreach ( $game->rounds[ 0 ]->creatures as $creature ) {
+                if ( isset( $grid[ $creature->locationx ][ $creature->locationy ] ) ) {
+                    $this->assertEquals( $grid[ $creature->locationx ][ $creature->locationy ], false, 'Only one creature must exist per coordinate' );
```
Comment (dionyziz) [inline tests/models/game.php]: assertFalse

File: models/game.php [hunk 38]
```diff
@@ -0,0 +1,75 @@
...
+            else {
+                $this->rounds = array();
+            }
+        }
+
+        protected function onBeforeCreate() {
+            $this->creaturesPerPlayer = rand( MIN_CREATURES, MAX_CREATURES );
+            $multiply = $this->creaturesPerPlayer * count( $this->users );
+            $this->width = rand( MIN_MULTIPLIER * $multiply + 1, MAX_MULTIPLIER * $multiply - 1 );
+            $this->height = rand( 3 * $multiply + 1, 4 * $multiply - 1 );
```
Comment (dionyziz) [inline models/game.php]: Use constant multipliers.

File: models/game.php [hunk 39]
```diff
@@ -0,0 +1,75 @@
...
+                        $x = rand( 0, $this->width - 1 );
+                        $y = rand( 0, $this->height - 1 );
+                        if ( !isset( $this->grid[ $x ][ $y ] ) ) {
+                            $creature->locationx = $x;
+                            $creature->locationy = $y;
+                            $this->grid[ $x ][ $y ] = $creature;
+                            break;
+                        }
+                    }
+                    $this->rounds[ 0 ]->creatures[ $id ] = $creature;
```
Comment (dionyziz) [inline models/game.php]: `$this->rounds[ 0 ]->creatures[] = $creature` and you don't need `$id`

**Pull request merged**